### PR TITLE
Reflection vs. source generation in System.Text.Json

### LINKED ISSUE #26040: Document known issues and semantic differences between the various JSON-serialization programming models
**Help us make content visible**

**Describe the new article**

With the new JSON source generator, we now have a few programming models for JSON (de)serialization:

- (De)serialization based on runtime reflection
- (De)serialization based on compile-time type metadata gathering (metadata source gen mode)
- Optimized serialization using Utf8JsonWriter directly

We should describe any feature and semantic differences between them, e.g. source-gen lack of support for init-only property deserialization (https://github.com/dotnet/runtime/issues/58770), lack of support for private members.

We should also describe feature differences in fast-path mode vs metadata serialization mode - https://github.com/dotnet/runtime/issues/51945#issuecomment-827795787.

---

cc @eiriktsarpalis, @steveharter, @tdykstra , @eerhardt, @ericstj, @jeffhandley, @jkotas, @stephentoub

### PR DESCRIPTION
**Pull request by tdykstra**
Fixes #26040 

File Type | File Name | Published Url
-- | -- | --
Content | docs/standard/serialization/system-text-json-overview.md | https://review.docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-overview?branch=pr-en-us-26559
Content | docs/standard/serialization/system-text-json-source-generation.md | https://review.docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-source-generation?branch=pr-en-us-26559
Content | docs/standard/serialization/system-text-json-source-generation-modes.md | https://review.docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-source-generation-modes?branch=pr-en-us-26559

### CODE REVIEW

**Review by gewarren — APPROVED** [review]
One small suggestion to consider because it took me a little while to figure out the columns.

File: docs/standard/serialization/system-text-json-source-generation-modes.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+:::zone pivot="dotnet-6-0"
+
+## Overview
+
+Choose reflection or source generation modes based on the following benefits that each one offers:
+
+| Benefit                                              | Reflection | Source generation:<br/>Metadata collection | Source generation:<br/>Serialization optimization |
+|------------------------------------------------------|------------|---------------------|----------------------------|
+| Simpler to code and debug.                           | ✔️        | ❌                  | ❌                        |
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-source-generation-modes.md:44]: In what ways do you find the reflection serializer simpler to code and debug?

File: docs/standard/serialization/system-text-json-source-generation-modes.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+| Supports non-public accessors.                       | ✔️        | ❌                  | ❌                        |
+| Supports init-only properties.                       | ✔️        | ❌                  | ❌                        |
+| Supports all available serialization customizations. | ✔️        | ❌                  | ❌                        |
+| Reduces start-up time.                               | ❌        | ✔️                  | ❌                        |
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-source-generation-modes.md:48]: Serialization optimization also reduces start-up time.

File: docs/standard/serialization/system-text-json-source-generation-modes.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+## Overview
+
+Choose reflection or source generation modes based on the following benefits that each one offers:
+
+| Benefit                                              | Reflection | Source generation:<br/>Metadata collection | Source generation:<br/>Serialization optimization |
+|------------------------------------------------------|------------|---------------------|----------------------------|
+| Simpler to code and debug.                           | ✔️        | ❌                  | ❌                        |
+| Supports non-public accessors.                       | ✔️        | ❌                  | ❌                        |
+| Supports init-only properties.                       | ✔️        | ❌                  | ❌                        |
+| Supports all available serialization customizations. | ✔️        | ❌                  | ❌                        |
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-source-generation-modes.md:47]: I believe metadata collection supports all customizations. Which ones did you find missing?

File: docs/standard/serialization/system-text-json-source-generation-modes.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+* How to access property getters and fields for serialization.
+* How to access a constructor, property setters, and fields for deserialization.
+* Information about which attributes have been used to customize serialization or deserialization.
+* Run-time configuration from <xref:System.Text.Json.JsonSerializerOptions>.
+
+This information is referred to as *metadata*.
+
+By default, `JsonSerializer` collects metadata at run time by using [reflection](../../csharp/programming-guide/concepts/reflection.md). Whenever `JsonSerializer` has to serialize or deserialize a type for the first time, it collects and caches this metadata. The metadata collection process takes time and uses memory.
+
+## Source generation - metadata collection mode
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-source-generation-modes.md:69]: Should this just be "Metadata collection mode"

File: docs/standard/serialization/system-text-json-source-generation-modes.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+You can use source generation to move the metadata collection process from run time to compile time. During compilation, the metadata is collected and source code files are generated. The generated source code files are automatically compiled as an integral part of the application. This compile-time metadata collection eliminates run-time metadata collection, which improves performance of both serialization and deserialization.
+
+The performance improvements provided by source generation can be substantial. For example, [test results](https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/#how-source-generation-provides-benefits) have shown up to 40% or more startup time reduction, private memory reduction, throughput speed increase (in serialization optimization mode), and app size reduction.
+
+## Source generation - known issues
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-source-generation-modes.md:75]: "Serialization optimization mode" should come before this section. Also should it just be "Known issues"?

File: docs/standard/serialization/system-text-json-source-generation-modes.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+## Source generation - metadata collection mode
+
+You can use source generation to move the metadata collection process from run time to compile time. During compilation, the metadata is collected and source code files are generated. The generated source code files are automatically compiled as an integral part of the application. This compile-time metadata collection eliminates run-time metadata collection, which improves performance of both serialization and deserialization.
+
+The performance improvements provided by source generation can be substantial. For example, [test results](https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/#how-source-generation-provides-benefits) have shown up to 40% or more startup time reduction, private memory reduction, throughput speed increase (in serialization optimization mode), and app size reduction.
```
Comment (layomia) [inline docs/standard/serialization/system-text-json-source-generation-modes.md:73]: I'd avoid quantifying the improvements here and just leave the pointer to the perf section of the blog.

**Review by layomia — COMMENTED** [review]
Just a few comments but LGTM

**Pull request merged**